packages/react-dom/src/shared/sanitizeURL.js

```diff
@@ -0,0 +1,47 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import invariant from 'shared/invariant';
+import warning from 'shared/warning';
+import ReactSharedInternals from 'shared/ReactSharedInternals';
+import {disableJavaScriptURLs} from 'shared/ReactFeatureFlags';
+
+let ReactDebugCurrentFrame = null;
+if (__DEV__) {
+  ReactDebugCurrentFrame = ReactSharedInternals.ReactDebugCurrentFrame;
+}
+
+// A javascript: URL can contain leading C0 control or \u0020 SPACE,
+// and any newline or tab are filtered out as if they're not part of the URL.
+// https://url.spec.whatwg.org/#url-parsing
+// Tab or newline are defined as \r\n\t:
+// https://infra.spec.whatwg.org/#ascii-tab-or-newline
+// A C0 control is a code point in the range \u0000 NULL to \u001F
+// INFORMATION SEPARATOR ONE, inclusive:
+// https://infra.spec.whatwg.org/#c0-control-or-space
+
+/* eslint-disable max-len */
+const isJavaScriptProtocol = /^[\u0000-\u001F ]*j[\r\n\t]*a[\r\n\t]*v[\r\n\t]*a[\r\n\t]*s[\r\n\t]*c[\r\n\t]*r[\r\n\t]*i[\r\n\t]*p[\r\n\t]*t[\r\n\t]*\:/i;
```

Comment:
Why not just grab everything before the first [:/?#] and then strip out everything that isn't a valid protocol character per [Std66](https://tools.ietf.org/html/std66) which I believes agrees with url.spec.whatwg:

    scheme        = ALPHA *( ALPHA / DIGIT / "+" / "-" / "." )

So something like

```js
function isJavascriptUrl(str) {
  return /^javascript$/i.test(  // Schemes are case insensitive
      /^[^:/#?]*/.exec(str)[0]  // Don't consider any part of any authority, path, query or fragment.
    .replace(/[^a-zA-Z0-9\+\-\.]+/g, '')  // Ignore non-scheme chars because Gareth Hayes
  );
}```

Reply:
Just intuition that this would be faster in the case where it's not a javascript URL, but I haven't actually benchmarked and I expect there to be a few alternatives benchmarked at some point. It's a fairly hot path since a lot of pages have many URLs.

Reply:
@sebmarkbage Fair enough.  https://gist.github.com/mikesamuel/1b76779ec5206e258829914e2b0dec27 uses a fast path check so is more benchmarkable version of the same algo with comments explaining the derivation.